### PR TITLE
Fixing Issue - 1655 - Dynamic form should not display any validation errors until the user clicks Save/Submit, or leaves the focus of a required control

### LINKED ISSUE #1655: Dynamic Form Loads with Validation Errors if no List Item Id is set
[ ] Enhancement

[X ] Bug

[ ] Question

#### Version

Please specify what version of the library you are using: [3.15.0]

#### Expected / Desired Behavior / Question
When loading a Dynamic Form with no list item id (new item), the form should not display any validation errors until the user clicks Save/Submit, or leaves the focus of a required control.

#### Observed Behavior
When loading a Dynamic Form with no list item id (new item) all of the required fields display validation errors by default.

#### Steps to Reproduce
Create A Dynamic Form with no list item id set, and link it to a list that has a required field.  You will see the errors load by default.

You can see this error reported by @sallaadithya in a sub comment for an existing bug here: https://github.com/pnp/sp-dev-fx-controls-react/issues/1586#issuecomment-1637459697

### PR DESCRIPTION
**Pull request by NishkalankBezawada**
| Q               | A
| --------------- | ---
| Bug fix?        | [x]
| New feature?    | [ ]
| New sample?      | [ ]
| Related issues?  | fixes #1655 

#### What's in this Pull Request?
Fixing Issue #1655 

#### Observation
When loading a Dynamic Form with no list item id (new item), the form should not display any validation errors until the user clicks Save/Submit, or leaves the focus of a required control.

#### Actual behavior
When loading a Dynamic Form with no list item id (new item) all of the required fields display validation errors by default.

#### Solution
In ```DynamicField.tsx``` component, individual field is set and the validation is done through the methods ```getRequiredErrorText()``` and ```getNumberErrorText()```

These methods currently are only checking with ```changedValue``` property, which will be always empty for a new item. An additional check is done on basis of ```listItemId``` assuming that the fact that for new item, this will be empty, we can achieve to not to show the validation errors on the form. 

Below is the current implementation of the code, 

```
  private getRequiredErrorText = (): string => {
    const {
      changedValue
    } = this.state;
    return (changedValue === undefined || changedValue === '' || changedValue === null || this.isEmptyArray(changedValue)) && this.props.required ? strings.DynamicFormRequiredErrorMessage : null;
  }
```

neccesary changes done as below, 

```
  private getRequiredErrorText = (): string => {
    const {
      changedValue,
      listItemId
    } = this.state;

    /**
     * checks if listItemId is not undefined, not an empty string, and not null.
     * If that condition is true happens when new item is loaded, it evaluates the nested ternary operator to determine the return value.
     * If the condition is false, it returns null.
     */

    return listItemId !== undefined && listItemId !== '' && listItemId !== null
      ? ((changedValue === undefined || changedValue === '' || changedValue === null || this.isEmptyArray(changedValue)) && this.props.required 
          ? strings.DynamicFormRequiredErrorMessage 
          : null)
      : null;

  }
```

#### Screenshots

Issue-1655

Thanks,
Nishkalank Bezawada

### CODE REVIEW

**Comment by michaelmaillot** [comment]
Hi @NishkalankBezawada, thanks for this PR!

I've reviewed it and couldn't make required field to be focused on during focus / blur & submit event for new items.

And after digging a bit about how the DynamicForm / DynamicField behave, it seems that the display of error messages for required fields is happening during the (re-)rendering process, which could happen for existing items but also for new ones.

I think that the display error messages for required fields should rely on the events which occur (blur & submit), whether you're working on an existing item or a new one.

**Review by michaelmaillot — CHANGES_REQUESTED** [review]

**Comment by michaelmaillot** [comment]
Sorry @NishkalankBezawada, I've added my review a moth ago but forgot to submit them... 🤦‍♂️

File: src/controls/dynamicForm/dynamicField/IDynamicFieldState.ts [hunk 1]
```diff
@@ -3,4 +3,5 @@ export interface IDynamicFieldState {
   * The options available to the listPicker
   */
   changedValue: any; // eslint-disable-line @typescript-eslint/no-explicit-any
+  listItemId:any;
```
[the remark this answers is not in the document]
Comment (NishkalankBezawada) [reply src/controls/dynamicForm/dynamicField/IDynamicFieldState.ts]: Hey @michaelmaillot ,

I have implemented as suggested. 

Thanks,
Nishkalank

File: src/controls/dynamicForm/dynamicField/DynamicField.tsx [hunk 2]
```diff
@@ -591,23 +592,38 @@ export class DynamicField extends React.Component<IDynamicFieldProps, IDynamicFi
...
+          ? strings.DynamicFormRequiredErrorMessage 
+          : null)
+      : null;
+
   }
 
   private getNumberErrorText = (): string => {
     const {
-      changedValue
+      changedValue, listItemId
```
[the remark this answers is not in the document]
Comment (NishkalankBezawada) [reply src/controls/dynamicForm/dynamicField/DynamicField.tsx]: Hey @michaelmaillot ,

I have implemented as suggested. 

Thanks,
Nishkalank

File: src/webparts/controlsTest/components/ControlsTest.tsx
[the remark this answers is not in the document]
Comment (NishkalankBezawada) [reply src/webparts/controlsTest/components/ControlsTest.tsx:1]: Hey @michaelmaillot ,

I have done neccesary changes as suggested. 

Thanks,
Nishkalank

**Comment by NishkalankBezawada** [comment]
Hey @michaelmaillot ,

Sorry for taking a lot of time. I have been  away for almost a month. Hope that you would review the changes. 

Thanks,
Nishkalank

**Comment by michaelmaillot** [comment]
Hi @NishkalankBezawada,

No worries! I'll have a look at your update and will come back to you asap!

**Comment by michaelmaillot** [comment]
Hi @NishkalankBezawada I currently can't test your updates since the ControlsTest webpart is not the latest one (there are still Toggle buttons displayed instead of Property Pane options).

See [here](https://github.com/pnp/sp-dev-fx-controls-react/blob/dev/src/webparts/controlsTest/components/ControlsTest.tsx) for reference.

Let me know if something's not clear regarding this.

**Comment by michaelmaillot** [comment]
Hi @NishkalankBezawada, are you still working on this PR? Do you need any help regarding your branch update?

**Comment by NishkalankBezawada** [comment]
> Hi @NishkalankBezawada, are you still working on this PR? Do you need any help regarding your branch update?

Hey @michaelmaillot, I was held up with other work and couldnt find time. I will do this evening, and keep you posted :)

//Nishkalank

**Comment by NishkalankBezawada** [comment]
> Hi @NishkalankBezawada, are you still working on this PR? Do you need any help regarding your branch update?

Unable to run gulp build or gulp serve on dev branch, Erroring out as below, 

image

**Pull request closed without merging**